### DISCUSSION TITLE
Query raw date comparison

### DISCUSSION THREAD

**Discussion by KopfSzmercen** [prisma/prisma#16088] [Q&A]
I've noticed a strange behaviour with my query raw prisma query that I want to perform. In my query, I want to take only matches that happened BEFORE a given date, however the query returns matches that started at the same time that I give to the query. However, when I hard-code the same date in the query, the match is not included, which is the behaviour I want. Any idea what is going on? I tried `  AND "Match"."startDateTime"::timestamptz < ${date}::timestamptz ` but with no successfull result. Tools I'm using:

"@prisma/client": "^4.4.0",
"prisma": "^4.4.0",
"@nestjs/core": "^9.0.0",
PostgreSQL 15.0 (Ubuntu 15.0-1.pgdg20.04+1) on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0, 64-bit

[Link to the image with log of query that returns also the match with the same date I provided to the query](https://i.postimg.cc/Qx9Lcd91/Screenshot-from-2022-11-01-11-00-56.png)

```
 //! THIS INCLUDES MATCH I DON'T WANT
        const date = new Date("2022-04-16T15:00:00.000Z")

        const count = await this.prisma.$queryRaw` 

        SELECT "Match"."id", SUM("Stats"."cornerKicks") as "sumOfCorners"

        FROM "Match"

        INNER JOIN "TeamMatchStatistics" as "Stats" ON  "Match".id =  "Stats"."matchId"
        INNER JOIN "Season" ON "Match"."seasonId" = "Season".id
     
        WHERE "Match"."awayTeamId" = 125
        
         -- HERE IS THE CONDITION 
        AND "Match"."startDateTime" < ${date}
        AND "Match"."statusDescriptionShort" IN ('FT', 'PEN', 'AET')

        AND "Season"."start" <= ${date}
        AND "Season"."end" >= ${date}
        
        GROUP BY "Match"."id"`

```
        //! THIS WORKS FINE
        
        const count = await this.prisma.$queryRaw` 

        SELECT "Match"."id", SUM("Stats"."cornerKicks") as "sumOfCorners"

        FROM "Match"

        INNER JOIN "TeamMatchStatistics" as "Stats" ON  "Match".id =  "Stats"."matchId"
        INNER JOIN "Season" ON "Match"."seasonId" = "Season".id
     
        WHERE "Match"."awayTeamId" = 125
        
         -- HERE IS THE CONDITION 
        AND "Match"."startDateTime" < '2022-04-16T15:00:00.000Z'
        AND "Match"."statusDescriptionShort" IN ('FT', 'PEN', 'AET')

        AND "Season"."start" <= ${date}
        AND "Season"."end" >= ${date}
        
        GROUP BY "Match"."id"`

**Comment by pantharshit00** [comment]
Can you please share the database structure here so that we can try this out? It would be even better if you can provide a script in a Git repository which can reproduce the issue so we can debug this more quickly.

**Reply by KopfSzmercen** [reply]
Here's the prisma model of Match
```
model Match {
    id            Int      @id @default(autoincrement())
    startDateTime DateTime

    refereeName            String?
    statusDescriptionLong  String
    statusDescriptionShort String
    statusElapsed          Int?

    winnerStatus WINNER_STATUS @default(UNKNOWN)

    goals Goal[]
    cards Card[]
    kicks Kick[]

    seasonId Int?
    venueId  Int?
    leagueId Int?

    homeTeamId Int?
    homeTeam   Team? @relation(name: "homeTeam", fields: [homeTeamId], references: [id])

    awayTeamId Int?
    awayTeam   Team? @relation(name: "awayTeam", fields: [awayTeamId], references: [id])

    updatedAt DateTime?
}
```
I created the same a similar query using 2 methods: queryRaw and findMany like this 
```
 const date = new Date("2022-04-16T15:00:00.000Z")

        const teamId = 125
        const a = await this.prisma.$queryRaw`
            SELECT "Match".id, "Match"."startDateTime"
            FROM "Match"
            WHERE "Match"."startDateTime" < ${date}
            AND "Match"."awayTeamId" = ${teamId} 
            OFFSET 0
        `

        console.log(a)

        const b = await this.prisma.match.findMany({
            where: {
                startDateTime: {
                    lt: date,
                },
                awayTeamId: teamId,
            },
            select: {
                id: true,
                startDateTime: true,
            },
        })

        console.log(b)
```
I compared the logs from prisma logger and logs on my Postgres container and that's how it looks
```
Prisma Logger - Query Raw
Prisma:query 
            SELECT "Match".id, "Match"."startDateTime"
            FROM "Match"
            WHERE "Match"."startDateTime" < $1
            AND "Match"."awayTeamId" = $2 
            OFFSET 0
        
Query: 
            SELECT "Match".id, "Match"."startDateTime"
            FROM "Match"
            WHERE "Match"."startDateTime" < $1
            AND "Match"."awayTeamId" = $2 
            OFFSET 0
        
Params: ["2022-04-16 15:00:00 UTC",125]
Duration: 2ms
```
```
Prisma Logger FindMany

prisma:query 
SELECT "public"."Match"."id", "public"."Match"."startDateTime" FROM "public"."Match" 
WHERE ("public"."Match"."startDateTime" < $1 AND "public"."Match"."awayTeamId" = $2) OFFSET $3
Query: 
SELECT "public"."Match"."id", "public"."Match"."startDateTime" FROM "public"."Match" 
WHERE ("public"."Match"."startDateTime" < $1 AND "public"."Match"."awayTeamId" = $2) OFFSET $3
Params: ["2022-04-16 15:00:00 UTC",125,0]
```
```
Postgres logs Find Many
2022-11-03 19:49:40.320 UTC [60] LOG:  execute s21: SELECT "public"."Match"."id", "public"."Match"."startDateTime"
 FROM "public"."Match" 
WHERE ("public"."Match"."startDateTime" < $1 AND "public"."Match"."awayTeamId" = $2) OFFSET $3

2022-11-03 19:49:40.320 UTC [60]
DETAIL:  parameters: $1 = '2022-04-16 15:00:00', $2 = '125', $3 = '0'
```

```
Postgres logs Query Raw
2022-11-03 19:49:40.318 UTC [60] LOG:  execute s20:
           SELECT "Match".id, "Match"."startDateTime"
           FROM "Match"
           WHERE "Match"."startDateTime" < $1
           AND "Match"."awayTeamId" = $2
           OFFSET 0
       
2022-11-03 19:49:40.318 UTC [60] DETAIL:  parameters: $1 = '2022-04-16 15:00:00+00', $2 = '125'
```
I noticed that the timezone of my db on which query raw returns me tha match that is starting at the same time as the given date is "Europe/Warsaw" so this causes the time shift. On my docker DB the timezone is UTC, so query raw works the same as find many. Nevertheless, I was confused by the logs of Prisma because I did not see the difference of date type "2022-04-16 15:00:0" vs  "2022-04-16 15:00:00+00" like in postgres logs.

**Reply by nurul3101** [reply]
@KopfSzmercen Hello 👋 
Can you try passing the date literal instead of the date object in the raw query, does it work as expected if the date literal is passed?

**Reply by pantharshit00** [reply]
👋 Hello @KopfSzmercen,  did you perhaps have a chance to check @nurul3101's recommendation? Let us know if you still have any questions!
 
 Or, if you've found a solution, we'd love to know how you solved this problem. Feel free to post it in this thread and "Mark as answer" if this discussion is now solved!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!

**Comment by bstchow** [comment]
@nurul3101 we ran into a similar issue wherein passing the date literal as a Postgres casted timestamp from the ISOString literal resolved the issue. However, it's extremely unclear why the version using an interpolated Date object does NOT work, since the Date object has the expected timezone.

In our case...

```
-- Unexpected behavior
t.created_at < ${new Date()}
```

```
-- Expected behavior
t.created_at < ${new Date().toISOString()}::timestamp
```
